Python/jit.c

```diff
@@ -146,256 +144,257 @@ set_bits(uint32_t *loc, uint8_t loc_start, uint64_t value, uint8_t value_start,
-                set_bits(loc32, 5, value, 2, 19);
-                continue;
-            case HoleKind_ARM64_RELOC_GOT_LOAD_PAGEOFF12:
-            case HoleKind_ARM64_RELOC_PAGEOFF12:
-            case HoleKind_IMAGE_REL_ARM64_PAGEOFFSET_12A:
-            case HoleKind_IMAGE_REL_ARM64_PAGEOFFSET_12L:
-            case HoleKind_R_AARCH64_ADD_ABS_LO12_NC:
-            case HoleKind_R_AARCH64_LD64_GOT_LO12_NC:
-                // 12-bit low part of an absolute address. Pairs nicely with
-                // ARM64_RELOC_GOT_LOAD_PAGE21 (above).
-                assert(IS_AARCH64_LDR_OR_STR(*loc32) || IS_AARCH64_ADD_OR_SUB(*loc32));
-                // There might be an implicit shift encoded in the instruction:
-                uint8_t shift = 0;
-                if (IS_AARCH64_LDR_OR_STR(*loc32)) {
-                    shift = (uint8_t)get_bits(*loc32, 30, 2);
-                    // If both of these are set, the shift is supposed to be 4.
-                    // That's pretty weird, and it's never actually been observed...
-                    assert(get_bits(*loc32, 23, 1) == 0 || get_bits(*loc32, 26, 1) == 0);
-                }
-                value = get_bits(value, 0, 12);
-                assert(get_bits(value, 0, shift) == 0);
-                set_bits(loc32, 10, value, shift, 12);
-                continue;
-        }
-        Py_UNREACHABLE();
+    uint32_t *loc32 = (uint32_t *)location;
+    // Check that we're not out of range of 32 unsigned bits:
+    assert(value < (1ULL << 32));
+    *loc32 = (uint32_t)value;
+}
+
+// 32-bit relative address.
+void
+patch_32r(unsigned char *location, uint64_t value)
+{
+    uint32_t *loc32 = (uint32_t *)location;
+    value -= (uintptr_t)location;
+    // Check that we're not out of range of 32 signed bits:
+    assert((int64_t)value >= -(1LL << 31));
+    assert((int64_t)value < (1LL << 31));
+    *loc32 = (uint32_t)value;
+}
+
+// 64-bit absolute address.
+void
+patch_64(unsigned char *location, uint64_t value)
+{
+    uint64_t *loc64 = (uint64_t *)location;
+    *loc64 = value;
+}
+
+// 12-bit low part of an absolute address. Pairs nicely with patch_aarch64_21r
+// (below).
+void
+patch_aarch64_12(unsigned char *location, uint64_t value)
+{
+    uint32_t *loc32 = (uint32_t *)location;
+    assert(IS_AARCH64_LDR_OR_STR(*loc32) || IS_AARCH64_ADD_OR_SUB(*loc32));
+    // There might be an implicit shift encoded in the instruction:
+    uint8_t shift = 0;
+    if (IS_AARCH64_LDR_OR_STR(*loc32)) {
+        shift = (uint8_t)get_bits(*loc32, 30, 2);
+        // If both of these are set, the shift is supposed to be 4.
+        // That's pretty weird, and it's never actually been observed...
+        assert(get_bits(*loc32, 23, 1) == 0 || get_bits(*loc32, 26, 1) == 0);
     }
+    value = get_bits(value, 0, 12);
+    assert(get_bits(value, 0, shift) == 0);
+    set_bits(loc32, 10, value, shift, 12);
 }
 
-static void
-copy_and_patch(unsigned char *base, const Stencil *stencil, uintptr_t patches[])
+void
```

Comment:
Maybe I'm missing something but I'm unclear why this function is required? Could we not just call `patch_aarch64_12` directly? Is this just a nice abstraction or are you future-proofing against something?

Reply:
See my other comment.